File: lib/domain/interaction_rule/ports/interaction_rule_destroy_output_port.rb
# frozen_string_literal: true

module Domain
  module InteractionRule
    module Ports
      class InteractionRuleDestroyOutputPort
        def on_success(destroy_output_dto)
          raise NotImplementedError, "Subclasses must implement on_success"
        end

        def on_failure(error_dto)
          raise NotImplementedError, "Subclasses must implement on_failure"
        end
      end
    end
  end
end

File: crates/agrr-domain/src/interaction_rule/ports/interaction_rule_destroy_output_port.rs
use crate::interaction_rule::dtos::InteractionRuleDestroyOutput;
use crate::shared::dtos::error::Error;
use crate::shared::policies::policy_permission_denied::PolicyPermissionDenied;

/// Ruby: `Domain::InteractionRule::Ports::InteractionRuleDestroyOutputPort`
pub trait InteractionRuleDestroyOutputPort {
    fn on_success(&mut self, dto: InteractionRuleDestroyOutput);
    fn on_failure(&mut self, error: DestroyFailure);
}

#[derive(Debug, Clone, PartialEq)]
pub enum DestroyFailure {
    Policy(PolicyPermissionDenied),
    Error(Error),
}
